WIP: Change to cryptonite

Vlix opened the pull request:
Started the switch to more than just `scrypt`.

- Changed over from `scrypt` package to the `cryptonite` package
- Added a `Symbol` parameter to the `PassHash` to keep it obvious which algorithm was used
- Added tests to check the hash is the same as from the `scrypt` package.
- Added functions to tweak the parameters for hashing

Vlix commented:
I'll be implementing the `bcrypt` functions somewhere this weekend hopefully.

Note to self: also update `password` version to 2.0.0.0 and add to ChangeLogs

cdepillabout commented:
I think this is looking really good.

I did a brief read-through and the only thing I really noticed is the phantom parameter.  It'd probably be easier to just declare an empty data type like

```haskell
data Scrypt
```

and use it like the following:

```haskell
newtype PassHash a = PassHash
  { unPassHash :: Text
  } deriving (Eq, Ord, Read, Show)

hashPass :: MonadIO m => Pass -> m (PassHash Scrypt)
hashPass = hashPassWithParams defaultParams
```

cdepillabout commented:
Also, if you're interested, I'd like to give you commit rights to this repo (as well as the package on Hackage) to make it easier for you to make changes like this.

It would also be helpful for you (and others?) if I were to ever go AWOL, especially if you're thinking of using this at work.

Vlix commented:
> I think this is looking really good.
> 
> I did a brief read-through and the only thing I really noticed is the phantom parameter. It'd probably be easier to just declare an empty data type like
> 
> ```haskell
> data Scrypt
> ```
> 
> and use it like the following:
> 
> ```haskell
> newtype PassHash a = PassHash
>   { unPassHash :: Text
>   } deriving (Eq, Ord, Read, Show)
> 
> hashPass :: MonadIO m => Pass -> m (PassHash Scrypt)
> hashPass = hashPassWithParams defaultParams
> ```

Hmmm, true, better to not force pragmas on users.



> Also, if you're interested, I'd like to give you commit rights to this repo (as well as the package on Hackage) to make it easier for you to make changes like this.
> 
> It would also be helpful for you (and others?) if I were to ever go AWOL, especially if you're thinking of using this at work.

Might be a good idea. Just to be sure. Not that I'll probably do a lot here, but who knows.

Vlix commented:
Also @cdepillabout , have you any idea why this happens in the `cabal` travis checks?

```
src/Data/Password/Scrypt.hs:35:1: error:
    Ambiguous module name ‘Data.ByteString.Base64’:
      it was found in multiple packages:
      base64-0.4.0 base64-bytestring-1.0.0.3
   |
35 | import qualified Data.ByteString.Base64 as Base64
```

cdepillabout commented:
> any idea why this happens in the cabal travis checks

I think it is due to doctest weirdness (basically, just not running in the right environment).  The easiest way to fix it is to just upgrade to using cabal-doctest.

For now you can ignore these failing tests.  We can update to cabal-doctest in a future PR.

Vlix commented:
Ok, I'll ignore the cabal tests.

Also, I'm not a big fan of unnecessary abbreviation (like with `PassCheckSucc`) so I was thinking of changing all mentions of `Pass` to `Password`. Like `hashPass` -> `hashPassword`, `PassHash` -> `PasswordHash`, etc. Do you have any objection to this?
(this is going to be a big rework of the library anyway, so if there's ever a time to do so it'd be now)

cdepillabout commented:
> I'm not a big fan of unnecessary abbreviation (like with PassCheckSucc) so I was thinking of changing all mentions of Pass to Password. Do you have any objection to this?

In general, I feel the same way.

In this specific instance, I was thinking that `Pass` is usually immediately obvious even when abbreviated.

However, I don't feel that strongly either way.  And most apps don't have _huge_ amounts of code dealing with passwords, so using the full word "password" shouldn't have a huge affect on how verbose the code feels.

So please feel free to change it.

cdepillabout commented:
I've just given you commit access here on github.

If you let me know your username on Hackage, I will give you access there as well.

Vlix commented:
> I've just given you commit access here on github.
> 
> If you let me know your username on Hackage, I will give you access there as well.

Thanks, I'll DM you my Hackage username on twitter. My twitter handle is `@FPaulusma`

Vlix commented:
Hey @cdepillabout , I'm not very familiar with `cabal`, so it'd be much appreciated if you could maybe check what's going on with the cabal failures. I've tried some tweaks to the `.travis.yml`, but there's always something poppin' up.

cdepillabout commented:
@Vlix Thanks for continuing work on this.

I'm happy if you just want to ignore the cabal errors for now.  As far as I can tell, they don't look like serious errors (something that would prevent cabal users from using this library), but instead just CI problems.

I've created a branch where I will try to fix the CI errors: https://github.com/cdepillabout/password/pull/10

cdepillabout commented:
@Vlix I was able to get CI passing again in https://github.com/cdepillabout/password/pull/10 (which is based on this PR).

If you want to get CI passing here in this PR, can you pull in https://github.com/cdepillabout/password/pull/10?

Vlix commented:
I'll do that next time I pick this up.

Vlix commented:
For future reference:
https://hashcat.net/wiki/doku.php?id=example_hashes

Vlix commented:
Ok, I feel like I've made pretty much all the code changes I wanted to. Hope this builds ok too.
I still would like to add a benchmarking suite to find better informed `defaultParams`, since I've been using the test suite for a good indication, but it does lots of parallel running etc, which skews the time indications; it's just not what you want to use to choose a default for the regular non-crypto savvy person.

Vlix commented:
Ok, https://travis-ci.org/github/cdepillabout/password/jobs/679539208 this errored... for SOME reason. Any way to just rebuild that one? Or do we have to make another commit to trigger a rebuild?

cdepillabout commented:
> Ok, I feel like I've made pretty much all the code changes I wanted to.

Great!  Want me to go ahead and do a final review, and then we can merge this in?

> Any way to just rebuild that one? Or do we have to make another commit to trigger a rebuild?

In the Travis CI UI, there is a rebuild button in the upper right hand corner.  However, I don't know if you will have permission to actually click it.  The easiest thing might be to just push a new commit.

I have added you as a collaborator on this repo, so I was hoping that would give you permissions on Travis CI, but maybe that doesn't automatically happen?  If this ever becomes a problem in the future, we could just switch this repo to be it's own organization here on Github, and then we'd hopefully both have access on Travis CI.

Vlix commented:
Yes, please. I'm comfortable with how it is now. The `defaultParam` tweaks after benchmarking is added will just be a minor version increase, since it will not break anything.

Vlix commented:
Small reminder to myself: might also be a good idea to update all the `README.md` files.

cdepillabout commented on password-instances/src/Data/Password/Instances.hs:
```diff
@@ -14,90 +14,79 @@ Maintainer  : cdep.illabout@gmail.com
+-- >>> let hashedPassword = hashPasswordWithSalt defaultParams salt pass
 -- >>> toPersistValue hashedPassword
--- PersistText "14|8|1|YWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXowMTIzNDU=|nENDaqWBmPKapAqQ3//H0iBImweGjoTqn5SvBS8Mc9FPFbzq6w65maYPZaO+SPamVZRXQjARQ8Y+5rhuDhjIhw=="
+-- PersistText "16|8|1|YWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXowMTIzNDU=|BH0oidcU/4Ec7Co4EM+LQ6xp39//MnOUhqmNeOnOz/nl4JHNXEJBw5dPdi3wTStYr+e1SmJkzHJrMvUJYNxK1w=="
```
Note to self: Add an item to the changelog that says the hash produced by `hashPasswordWithSalt` has changed in this new version.

Vlix replied:
Well, `hashPasswordWithSalt` is completely different, since it takes parameters now. And the hash is only different because the rounds are higher, if you change `defaultParams` to `defaultParams{ scryptRounds = 14 }`, it'll produce the same hash.

cdepillabout replied:
@Vlix I was mostly just worried about people that are currently using `password` in production.  They might want an easy way to produce the password hashes in the same format.

Of course, I haven't actually heard from anyone using `password` in production, so maybe it doesn't matter at all.

Vlix replied:
The format is not changed. I've even got tests that verify it in `test/Scrypt.hs`

We can set the default cost back to 14 until we get benchmarks that provide a more sensible default value (like I was planning for `bcrypt` anyway) so that the hashes stay the same if using `defaultParams`?

cdepillabout replied:
> The format is not changed.

Ah, I see, so it is just the cost that has been changed from 14 to 16.

Okay, how about this:

We leave it as you currently have it in the PR, but you add a small note to the `CHANGELOG` that tells people to set the `cost` argument in the bcrypt `defaultParams` if they want the same functionality from the old version of `password`?

Although, like I said above, I don't know if there is really anyone using `password`, so if you don't want to do this, I'd be fine with going ahead and doing a release anyway.

I don't want my small nitpicks to hold up a new release.

Vlix replied:
Well, it's a fair point, and since I increased the rounds just because the python library has 16 as default, and it seemed fast enough in the test cases, I think it's better to just keep it at 14 and do all the default tweaking after the benchmarking is done.
This makes it more in line with the thought of the other defaults as well.

cdepillabout replied:
Okay, that sounds good to me.

Vlix replied:
Updated

cdepillabout commented on password-instances/src/Data/Password/Instances.hs:
```diff
@@ -14,90 +14,79 @@ Maintainer  : cdep.illabout@gmail.com
+instance PersistField (PasswordHash a) where
+  toPersistValue (PasswordHash hpw) = PersistText hpw
+  fromPersistValue (PersistText txt) = Right $ PasswordHash txt
+  fromPersistValue (PersistByteString bs) = Right $ PasswordHash $ TE.decodeUtf8 bs
```
I don't think it is a big issue, but we could probably take this line out?

```suggestion
```

If we decide to leave this line in, we should probably mention it in the documentation.

Vlix replied:
I would not take this out. If someone used ByteString (SQL type: `bytea`) as type of the column for the hashes before using this package (or maybe even before using Haskell) it will fail to parse, which is unnecessary.

cdepillabout replied:
Okay, that's a good point.  I agree with keeping it in then.

cdepillabout commented on password/password.cabal:
```diff
@@ -26,35 +32,42 @@ library
+      Data.Password.Scrypt
   other-modules:
       Paths_password
+      Data.Password.Internal
```
I'm a fan of exposing these `Internal` modules for power users.

Although, we could always just wait until someone explicitly requests it.

Vlix replied:
Everything from `Data.Password.Internal` is re-exported by `Data.Password`, except the data constructor of `Password`, but you can still get to the `Text` inside using the `unsafeShowPassword` functions.

The other utility functions are irrelevant for this package (`showT readT toBytes fromBytes`)

Vlix replied:
Ok, let's do that.

cdepillabout commented on password/src/Data/Password.hs:
```diff
@@ -8,17 +6,41 @@ Maintainer  : cdep.illabout@gmail.com
+
+== Algorithms
+
+Generally, the most "secure" algorithm is believed to be @Argon2@, then @scrypt@,
+then @bcrypt@, and lastly @PBKDF2@. @bcrypt@ and @PBKDF2@ are the most established
+algorithms, so they have been tried and tested, though they both lack a memory cost,
+and therefore have a greater vulnerability to specialized hardware attacks.
+
+When choosing an algorithm, and you have no idea which to pick, just go for @bcrypt@
+if your password does not need the highest security possible. It's still a fine way
+for hashing passwords, and the cost is easily adjustable if needed. If your needs
+do require stronger protection, you should find someone who can advise you on this
+topic. (And if you're already knowledgeable enough, you know what to do)
```
Great explanation here!

cdepillabout commented on password/src/Data/Password/Argon2.hs:
I think it is nice to have this explicitly in the haddocks.

```suggestion
--
-- >>> defaultParams
-- Argon2Params {argon2Salt = 16, argon2Variant = Argon2id, argon2Version = Version13, argon2MemoryCost = 65536, argon2TimeCost = 2, argon2Parallelism = 1, argon2OutputLength = 32}
--
```

cdepillabout replied:
It'd be nice to have this in the other modules with params as well.

Vlix replied:
That's a good point.

Vlix replied:
Updated

cdepillabout commented on stack.yaml:
```diff
@@ -39,7 +39,8 @@ packages:
 # using the same syntax as the packages field.
 # (e.g., acme-missiles-0.3)
-# extra-deps: []
+extra-deps:
+  - base64-0.3.1.1
```
I'd like to update this to lts-15, and get rid of these extra-deps.  Ideally I'd like to keep `password` in stackage.

cdepillabout replied:
I just tried doing the following and everything still builds with `stack`:

```diff
diff --git a/stack.yaml b/stack.yaml
index 3e0249a..cf81507 100644
--- a/stack.yaml
+++ b/stack.yaml
@@ -17,7 +17,7 @@
 #
 # resolver: ./custom-snapshot.yaml
 # resolver: https://example.com/snapshots/2018-01-01.yaml
-resolver: lts-14.20
+resolver: lts-15.10
 
 # User packages to be built.
 # Various formats can be used as shown in the example below.
@@ -39,8 +39,8 @@ packages:
 # Dependency packages to be pulled from upstream that are not in the resolver
 # using the same syntax as the packages field.
 # (e.g., acme-missiles-0.3)
-extra-deps:
-  - base64-0.3.1.1
+extra-deps: []
 
 # Override default flag values for local packages and extra-deps
 # flags: {}
```

Vlix replied:
Yeah, `base64` was added to stackage since `lts-15.0`

Vlix replied:
Updated

Vlix commented:
Yeah, just the added documentation for all the `defaultParams` is something I'd like to get in there. I'm also thinking of removing `unsafeShowPasswordText` and make `unsafeShowPassword :: Password -> Text`, since anyone can change it to `String` if they want with `Data.Text.unpack`.

cdepillabout commented:
> I'm also thinking of removing unsafeShowPasswordText and make unsafeShowPassword :: Password -> Text

Yeah, that sounds good to me!

Vlix commented:
Ok, I've implemented all the changes. Also added instances in `password-instances` with custom type errors. It looks [\_\_like_this\_\_](https://i.ibb.co/Ks5LwrN/Screenshot-from-2020-04-30-01-18-56.png)

cdepillabout commented:
@Vlix Okay, this looks good to me.  Do you want to merge this in and make a release to Hackage?

You probably don't need it, but here is my checklist for making a release to Hackage:

https://functor.tokyo/blog/2018-07-16-release-haskell-packages-to-hackage

Vlix commented:
Sure, I can merge this in. #10 can be closed then, I think.

Your checklist doesn't seem to include uploading the documentation and doesn't use the upload candidate functionality. I've only uploaded to hackage with `cabal`, so I wonder if `stack upload` just immediately uploads it for real, or does it start an interactive menu or something?

cdepillabout commented:
> Your checklist doesn't seem to include uploading the documentation and doesn't use the upload candidate functionality. 

I normally tend to let Hackage build the docs for me, unless it can't, in which case I upload the documentation.

Running `stack upload password` and then `stack upload password-instances` should upload the two packages, and then Hackage will hopefully build the documentation for them.

However, if you're used to doing development with `cabal`, then using cabal to upload the packages is completely okay with me!

Vlix commented:
I took a last pass over the documentation for links etc. I'm also not sure what the proper etiquette is about the "Copyright" field at the top of the modules. Do I add my name there if I add (significantly) to the module? What if I just shuffle functions around? Should just the person be named that created the library?
I'm not sure what the regular procedures are for open-source projects. 🤔

cdepillabout commented:
> I'm also not sure what the proper etiquette is about the "Copyright" field at the top of the modules. Do I add my name there if I add (significantly) to the module? What if I just shuffle functions around? Should just the person be named that created the library?

I'm also not sure about this.  There is also the `LICENSE` file (and maybe a `copyright` section in the `.cabal` file?).

I'd be happy if you wanted to add your name in all these places, as well as the author/maintainer section in the `.cabal` files, especially if you'll help me keep an eye on this library in the future.

Feel free to add your name, merge this in, and cut a new release.

Vlix commented:
I don't mind maintaining the package as well. I've added my name in the appropriate places and will merge + upload later today.

cdepillabout commented:
@Vlix I'd like to write a blog post about this update to the `password` library and post it to /r/haskell and twitter so more people are aware of it.

However, I imagine you'd be able to write a better post than me, given that you did all of the rewrite.  Do you have a blog and would be willing to create a post about this?

If not, no worries, I'll throw something together.

Vlix commented:
@cdepillabout I don't have a blog, no. So you can go ahead and write something up. Maybe a good idea to let me do a read over and adjust it here and there if need be.
